torch/csrc/autograd/python_function.cpp

```diff
@@ -366,100 +349,82 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
     self->output_info.reserve(num_outputs);
   }
 
-  auto shared_pairs = _parse_shared_pairs(self, t2var);
+  std::unordered_set<PyObject*> inputs;
+  int num_inputs = PyTuple_GET_SIZE(inputs_tuple);
+  for (int i = 0; i < num_inputs; i++) {
+    inputs.emplace(PyTuple_GET_ITEM(inputs_tuple, i));
+  }
+
   auto non_differentiable = _parse_non_differentiable(self);
+  auto dirty_inputs = _parse_dirty_inputs(self);
 
-  // Given an output tensor, find the input Variable with which it shares storage
-  auto get_shared_base = [&](PyObject* tensor) -> Variable {
-    auto input_it = t2var.find(tensor);
-    if (input_it != t2var.end()) {
-      // If the output is an input treat that as the base
-      return input_it->second->cdata;
+  auto as_variable = [&](PyObject* obj, int i) -> Variable {
+    if (THPVariable_Check(obj)) {
+      return ((THPVariable*)obj)->cdata;
     }
-    auto it = shared_pairs.find(tensor);
-    if (it != shared_pairs.end()) {
-      // It's explicitly marked as shared via mark_shared_storage
-      return it->second->cdata;
+    if (THPModule_isTensor(obj)) {
+      // temporarily wrap tensors as variables until the classes are merged
+      return make_variable(createTensor(obj));
     }
-    return Variable();
+    throw TypeError("%s.forward: expected Variable (got %s) for return value %d",
+        Py_TYPE(self)->tp_name, Py_TYPE(obj)->tp_name, i);
   };
 
   // Wraps an output Tensor in a Variable or returns the previous wrapper in
   // the case of in-place modification.
-  auto wrap_output = [&](at::Tensor data, Variable prev, int output_nr, bool is_modified, bool is_non_differentiable) -> Variable {
+  auto set_history = [&](Variable& var, int output_nr, bool is_input, bool is_modified,
+                         bool is_non_differentiable) {
     if (is_non_differentiable) {
-      return make_variable(std::move(data));
-    }
-    if (!prev.defined()) {
-      return make_variable(std::move(data), output_nr, cdata);
-    }
-    if (is_modified) {
-      if (prev.is_leaf() && prev.requires_grad()) {
+      var.detach_();
+    } else if (is_modified) {
+      if (var.is_leaf() && var.requires_grad()) {
         throw std::runtime_error("a leaf Variable that requires grad has been used in an in-place operation.");
       }
       // If the input was modified, transplant the grad_fn in the graph:
       // grad_fn <- variable <- self  ==>  grad_fn <- self <- variable
-      prev.get()->grad.reset();
-      prev.get()->hooks.clear();
-      if (auto grad_acc_fn = prev.get()->grad_accumulator.lock()) {
+      var.get()->grad.reset();
+      var.get()->hooks.clear();
+      if (auto grad_acc_fn = var.get()->grad_accumulator.lock()) {
         auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
         grad_acc->variable.reset();
       }
       if (cdata) {
-        prev.rebase_history(output_nr, cdata);
+        var.rebase_history(output_nr, cdata);
```

Comment:
`rebase_history` has a special implementation for views, right?

Reply:
Yes. For views it will modify the `grad_fn` on the base (using `cdata` here). The `grad_fn` on the view reverses the viewing operation.

On non-views, it just sets `var.output_nr=output_nr` and `var.grad_fn=cdata`